test/hpke_test.c

```diff
@@ -0,0 +1,1804 @@
+    unsigned char rexp[32];
+    unsigned char rexp2[32];
+    unsigned char plain[] = "quick brown fox";
+    size_t plainlen = sizeof(plain);
+    unsigned char enc[OSSL_HPKE_TSTSIZE];
+    size_t enclen = sizeof(enc);
+    unsigned char cipher[OSSL_HPKE_TSTSIZE];
+    size_t cipherlen = sizeof(cipher);
+    unsigned char clear[OSSL_HPKE_TSTSIZE];
+    size_t clearlen = sizeof(clear);
+    char *estr = "foo";
+
+    if (!TEST_true(OSSL_HPKE_keygen(hpke_suite, pub, &publen, &privp,
+                                    NULL, 0, testctx, NULL)))
+        goto end;
+    if (!TEST_ptr(ctx = OSSL_HPKE_CTX_new(hpke_mode, hpke_suite,
+                                          testctx, NULL)))
+        goto end;
+    /* a few error cases 1st */
+    if (!TEST_false(OSSL_HPKE_export(NULL, exp, sizeof(exp),
+                                     (unsigned char *)estr, strlen(estr))))
+        goto end;
+    /* ctx before encap should fail too */
+    if (!TEST_false(OSSL_HPKE_export(ctx, exp, sizeof(exp),
+                                     (unsigned char *)estr, strlen(estr))))
+        goto end;
+    if (!TEST_true(OSSL_HPKE_encap(ctx, enc, &enclen, pub, publen, NULL, 0)))
+        goto end;
+    if (!TEST_true(OSSL_HPKE_seal(ctx, cipher, &cipherlen, NULL, 0,
+                                  plain, plainlen)))
+        goto end;
+    /* now for real */
+    if (!TEST_true(OSSL_HPKE_export(ctx, exp, sizeof(exp),
+                                    (unsigned char *)estr, strlen(estr))))
+        goto end;
+    /* check a 2nd call with same input gives same output */
+    if (!TEST_true(OSSL_HPKE_export(ctx, exp2, sizeof(exp2),
+                                    (unsigned char *)estr, strlen(estr))))
+        goto end;
+    if (!TEST_true(TEST_mem_eq(exp, sizeof(exp), exp2, sizeof(exp2))))
+        goto end;
+    if (!TEST_ptr(rctx = OSSL_HPKE_CTX_new(hpke_mode, hpke_suite,
+                                           testctx, NULL)))
+        goto end;
+    if (!TEST_true(OSSL_HPKE_decap(rctx, enc, enclen, privp, NULL, 0)))
+        goto end;
+    if (!TEST_true(OSSL_HPKE_open(rctx, clear, &clearlen, NULL, 0,
+                                  cipher, cipherlen)))
+        goto end;
+    if (!TEST_true(OSSL_HPKE_export(rctx, rexp, sizeof(rexp),
+                                    (unsigned char *)estr, strlen(estr))))
+        goto end;
+    /* check a 2nd call with same input gives same output */
+    if (!TEST_true(OSSL_HPKE_export(rctx, rexp2, sizeof(rexp2),
+                                    (unsigned char *)estr, strlen(estr))))
+        goto end;
+    if (!TEST_true(TEST_mem_eq(rexp, sizeof(rexp), rexp2, sizeof(rexp2))))
+        goto end;
+    if (!TEST_true(TEST_mem_eq(exp, sizeof(exp), rexp, sizeof(rexp))))
+        goto end;
+    OSSL_HPKE_CTX_free(ctx);
+    OSSL_HPKE_CTX_free(rctx);
+    EVP_PKEY_free(privp);
+    return 1;
+end:
+    OSSL_HPKE_CTX_free(ctx);
+    OSSL_HPKE_CTX_free(rctx);
+    EVP_PKEY_free(privp);
```

Comment:
Better would be to have one common cleanup block and just return a "testresult" value at the end.

Reply:
sure